swarm/api/config.go

```diff
@@ -46,101 +43,63 @@ type Config struct {
-	err = os.MkdirAll(dirpath, os.ModePerm)
-	if err != nil {
-		return
-	}
-	confpath := filepath.Join(dirpath, "config.json")
-	var data []byte
-	pubkey := crypto.FromECDSAPub(&prvKey.PublicKey)
-	pubkeyhex := common.ToHex(pubkey)
-	keyhex := crypto.Keccak256Hash(pubkey).Hex()
+//create a default config with all parameters to set to defaults
+func NewDefaultConfig() (self *Config) {
 
 	self = &Config{
-		SyncParams:    network.NewSyncParams(dirpath),
-		HiveParams:    network.NewHiveParams(dirpath),
+		StoreParams:   storage.NewDefaultStoreParams(),
 		ChunkerParams: storage.NewChunkerParams(),
-		StoreParams:   storage.NewStoreParams(dirpath),
+		HiveParams:    network.NewDefaultHiveParams(),
+		SyncParams:    network.NewDefaultSyncParams(),
+		Swap:          swap.NewDefaultSwapParams(),
 		ListenAddr:    DefaultHTTPListenAddr,
 		Port:          DefaultHTTPPort,
-		Path:          dirpath,
-		Swap:          swap.DefaultSwapParams(contract, prvKey),
-		PublicKey:     pubkeyhex,
-		BzzKey:        keyhex,
+		Path:          node.DefaultDataDir(),
+		EnsApi:        node.DefaultIPCEndpoint("geth"),
 		EnsRoot:       ens.TestNetAddress,
-		NetworkId:     networkId,
-	}
-	data, err = ioutil.ReadFile(confpath)
-
-	// if not set in function param, then set default for swarm network, will be overwritten by config file if present
-	if networkId == 0 {
-		self.NetworkId = network.NetworkId
+		NetworkId:     network.NetworkId,
+		SwapEnabled:   false,
+		SyncEnabled:   true,
+		SwapApi:       "",
 	}
 
-	if err != nil {
-		if !os.IsNotExist(err) {
-			return
-		}
-
-		// file does not exist
-		// write out config file
-		err = self.Save()
-		if err != nil {
-			err = fmt.Errorf("error writing config: %v", err)
-		}
-		return
-	}
+	return
+}
 
-	// file exists, deserialise
-	err = json.Unmarshal(data, self)
-	if err != nil {
-		return nil, fmt.Errorf("unable to parse config: %v", err)
-	}
-	// check public key
-	if pubkeyhex != self.PublicKey {
-		return nil, fmt.Errorf("public key does not match the one in the config file %v != %v", pubkeyhex, self.PublicKey)
-	}
-	if keyhex != self.BzzKey {
-		return nil, fmt.Errorf("bzz key does not match the one in the config file %v != %v", keyhex, self.BzzKey)
-	}
+//some config params need to be initialized after the complete
+//config building phase is completed (e.g. due to overriding flags)
+func (self *Config) Init(prvKey *ecdsa.PrivateKey) {
 
-	// if set in function param, replace id set from config file
-	if networkId != 0 {
-		self.NetworkId = networkId
-	}
+	pubkey := crypto.FromECDSAPub(&prvKey.PublicKey)
+	pubkeyhex := common.ToHex(pubkey)
+	keyhex := crypto.Keccak256Hash(pubkey).Hex()
 
-	self.Swap.SetKey(prvKey)
+	self.PublicKey = pubkeyhex
+	self.BzzKey = keyhex
 
-	if (self.EnsRoot == common.Address{}) {
-		self.EnsRoot = ens.TestNetAddress
+	//create the directory if it doesn't exist
+	if _, err := os.Stat(self.Path); os.IsNotExist(err) {
+		os.MkdirAll(self.Path, os.ModePerm)
 	}
 
```

Comment:
Should this create the `<datadir>/swarm/bzz-<key>` datadir? Now it's set to the same as the node's?

Reply:
Only if it doesn't exist? 

It would actually create `--datadir` if it wouldn't exist - which should not happen at this point, but added this just to be safe.

I actually added this last minute, after some running tests were producing errors (which then turned out not being related to the datadir). I wonder if this is needed indeed?

Reply:
well it could happen if we never ran `geth` before and are running `swarm --ens-api ''`? 

My point was where is the `swarm/bzz-xxxx` subdir created? I missed it maybe?

Reply:
This is a spot-on remark! Made me realize quite a few things! Thanks @nolash ! (And `bzz-xxxx` wasn't even being created at all!)